File: stack,queue/올바른괄호.js
class Stack {
  constructor() {
    this.items = [];
  }
  push(item) {
    this.items.push(item);
  }
  pop() {
    this.items.pop();
  }
  empty() {
    return this.items.length === 0;
  }
}

function solution(s) {
  var answer = true;

  const stack = new Stack();

  [...s].forEach((item) => {
    if (item === "(") {
      stack.push(item);
    } else {
      if (stack.empty()) {
        answer = false;
        return;
      }
      stack.pop();
    }
  });

  if (!stack.empty()) return false;

  return answer;
}

File: programmers/Lv2/올바른괄호.cpp
#include <string>
#include <iostream>
#include <stack>

using namespace std;

bool solution(string s)
{
    bool answer = true;
    stack<int> stack;

    for(char c : s) {
        // "(" 일 때
        if(c == '(') {
            stack.push(c);
        }
        // ")" 일 때
        else if(c == ')'){
            if(!stack.empty()) stack.pop();
            else {
                answer = false;
                break;
            }
        }
    }
    
    if(!stack.empty()) answer = false;

    return answer;
}